sled-storage/src/manager.rs

```diff
@@ -0,0 +1,969 @@
+
+        // Disks that should be requeued.
+        let mut saved = HashSet::new();
+        let queued = std::mem::take(&mut self.queued_u2_drives);
+        let mut iter = queued.into_iter();
+        while let Some(disk) = iter.next() {
+            if self.state == StorageManagerState::QueuingDisks {
+                // We hit a transient error in a prior iteration.
+                saved.insert(disk);
+            } else {
+                // Try ot add the disk. If there was a transient error the disk will
+                // have been requeued. If there was a permanent error, it will have been
+                // dropped. If there is an another unexpected error, we will handle it and
+                // requeue ourselves.
+                match self.add_u2_disk(disk.clone()).await {
+                    Err(err) => {
+                        warn!(
+                            self.log,
+                            "Potentially transient error: {err}: requeuing disk";
+                            "disk_id" => ?disk.identity()
+                        );
+                        saved.insert(disk);
+                    }
+                    Ok(true) => send_updates = true,
+                    Ok(false) => (),
+                }
+            }
+        }
+        // Merge any requeued disks from transient errors with saved disks here
+        self.queued_u2_drives.extend(saved);
+        send_updates
+    }
+
+    // Add a disk to `StorageResources` if it is new and return Ok(true) if so
+    async fn add_disk(&mut self, raw_disk: RawDisk) -> Result<bool, Error> {
+        match raw_disk.variant() {
+            DiskVariant::U2 => self.add_u2_disk(raw_disk).await,
+            DiskVariant::M2 => self.add_m2_disk(raw_disk).await,
+        }
+    }
+
+    // Add a U.2 disk to [`StorageResources`] or queue it to be added later
+    async fn add_u2_disk(&mut self, raw_disk: RawDisk) -> Result<bool, Error> {
+        if self.state != StorageManagerState::Normal {
+            self.queued_u2_drives.insert(raw_disk);
+            return Ok(false);
+        }
+
+        match Disk::new(&self.log, raw_disk.clone(), Some(&self.key_requester))
+            .await
+        {
+            Ok(disk) => self.resources.insert_disk(disk),
+            Err(err @ DiskError::Dataset(DatasetError::KeyManager(_))) => {
+                warn!(
+                    self.log,
+                    "Transient error: {err}: queuing disk";
+                    "disk_id" => ?raw_disk.identity()
+                );
+                self.queued_u2_drives.insert(raw_disk);
+                self.state = StorageManagerState::QueuingDisks;
+                Ok(false)
+            }
+            Err(err) => {
+                error!(
+                    self.log,
+                    "Persistent error: {err}: not queueing disk";
+                    "disk_id" => ?raw_disk.identity()
+                );
+                Ok(false)
+            }
```

Comment:
The question is "who" has no idea? The caller in this case is the task manager itself which received an update. All it does in the case of error when this internal method returns is to log an error. And when a success is returned it is determining whether to broadcast an update based on a boolean.

 However, the local method itself knows better whether this is a transient error or permanent error and is in a place to log it better than the caller. If we don't do it this way, all that will happen is that we will end up duplicating this behavior in the caller. In either case, an observer external to the task will not know the difference or be able to tell about a permanent error. While we could "broadcast" the error, and we may want to in the future, there's nothing to do other than log it currently as we don't have any sort of fault alerting yet.

Reply:
After thinking more about this, I agree with you @smklein. I'm going to change the code around to either log all errors in this method and not return them or always return an error when relevant. This actually helps a bit in making `add_queued_disks` cancellation safe.

Reply:
I cleaned this up using an enum instead of a bool for the return type and always returning any permanent errors in [11f1b36](https://github.com/oxidecomputer/omicron/pull/4332/commits/11f1b3695b4954e94ce40b9680e8e863b735c56f). I think it's much clearer now and no longer requires as detailed comments for the return type. Thanks for pointing this out!